### ISSUE TITLE
Error using a composite type as partitioning key

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#9405]
### Greenplum version or build
master, 6X_STABLE, 5X_STABLE.

### Step to reproduce the behavior

```
postgres=# create type employee_type as (empid int, empname text);
CREATE TYPE
postgres=# create table employee_table(timest date, user_id numeric(16,0) not null, tag1 char(5), emp employee_type)
postgres-#   distributed by (user_id)
postgres-#   partition by list(emp)
postgres-#   (partition part1 values('(1, ''foo'')'::employee_type), partition part2 values('(2, ''foo'')'::employee_type));
NOTICE:  CREATE TABLE will create partition "employee_table_1_prt_part1" for table "employee_table"
NOTICE:  CREATE TABLE will create partition "employee_table_1_prt_part2" for table "employee_table"
ERROR:  input of anonymous composite types is not implemented
```
I expected this to work without an error. 

I ran into it while debugging a pg_upgrade failure with a similar case with using a composite type as partitioning key. The above works, if you also create a custom operator and opclass
```
create type employee_type as (empid int, empname text);
create function emp_equal(employee_type, employee_type) returns boolean
  as 'select $1.empid = $2.empid;'
  language sql
  immutable
  returns null on null input;
create operator = (
        leftarg = employee_type,
        rightarg = employee_type,
        procedure = emp_equal
);
create operator class employee_incomplete_op_class default for type employee_type
  using btree as
  operator 3 =;
create table employee_table(timest date, user_id numeric(16,0) not null, tag1 char(5), emp employee_type)
  distributed by (user_id)
  partition by list(emp)
  (partition part1 values('(1, ''foo'')'::employee_type), partition part2 values('(2, ''foo'')'::employee_type));
```
However, pg_upgrade fails when trying to upgrade a database containing that:
```
$ ./bin/pg_upgrade --new-bindir=bin --old-bindir=bin --old-datadir=data-master --new-datadir=data-master2 --mode=dispatcher
Performing Consistency Checks
-----------------------------
Checking cluster versions                                   ok
Checking database user is the install user                  ok
Checking database connection settings                       ok
Checking for prepared transactions                          ok
Checking for reg* system OID user data types                ok
Checking for contrib/isn with bigint-passing mismatch       ok
Checking for online expansion status                        ok
Checking for external tables used in partitioning           ok
Checking for non-covering indexes on partitioned AO tables  ok
Checking for indexes on partitioned tables                  ok
Checking for orphaned TOAST relations                       ok
Checking array types derived from partitions                ok
Creating dump of global objects                             ok
Creating dump of database schemas
                                                            ok
Checking for presence of required libraries                 ok
Checking database user is the install user                  ok
Checking for prepared transactions                          ok

If pg_upgrade fails after this point, you must re-initdb the
new cluster before continuing.

Performing Upgrade
------------------
Analyzing all rows in the new cluster                       ok
Freezing all rows on the new cluster                        ok
Deleting files from new pg_clog                             ok
Copying old pg_clog to new server                           ok
Setting next transaction ID and epoch for new cluster       ok
Deleting files from new pg_multixact/offsets                ok
Copying old pg_multixact/offsets to new server              ok
Deleting files from new pg_multixact/members                ok
Copying old pg_multixact/members to new server              ok
Setting next multixact ID and offset for new cluster        ok
Resetting WAL archives                                      ok
Deleting files from new pg_distributedlog                   ok
Copying old pg_distributedlog to new server                 ok
Setting frozenxid and minmxid counters in new cluster       ok
Restoring global objects in the new cluster                 ok
Restoring database schemas in the new cluster
  postgres                                                  
*failure*

Consult the last few lines of "pg_upgrade_dump_13194.log" for
the probable cause of the failure.
Failure, exiting
```
`pg_upgrade_dump_13194.log`:
```
command:  PGOPTIONS='-c gp_session_role=utility' "bin/pg_restore" --host '/home/heikki/gpdb.master' --port 50432 --username 'heikki' --exit-on-error --binary-upgrade --verbose --dbname 'dbname=postgres' "pg_upgrade_dump_13194.custom" >> "pg_upgrade_dump_13194.log" 2>&1
pg_restore: connecting to database for restore
pg_restore: dropping SCHEMA public
pg_restore: creating pg_largeobject "pg_largeobject"
pg_restore: creating pg_largeobject_metadata "pg_largeobject_metadata"
pg_restore: creating COMMENT "DATABASE "postgres""
pg_restore: creating binary_upgrade "binary_upgrade"
pg_restore: creating SCHEMA "public"
pg_restore: creating COMMENT "SCHEMA "public""
pg_restore: creating TYPE "public.comptype"
pg_restore: creating TYPE "public.employee_type"
pg_restore: creating FUNCTION "public.comptype_equal("public"."comptype", "public"."comptype")"
pg_restore: creating FUNCTION "public.emp_equal("public"."employee_type", "public"."employee_type")"
pg_restore: creating OPERATOR "public.="
pg_restore: creating OPERATOR FAMILY "public.employee_incomplete_op_class"
pg_restore: creating OPERATOR CLASS "public.employee_incomplete_op_class"
pg_restore: creating TABLE "public.employee_table"
pg_restore: [archiver (db)] Error while PROCESSING TOC:
pg_restore: [archiver (db)] Error from TOC entry 281; 1259 24921 TABLE employee_table heikki
pg_restore: [archiver (db)] could not execute query: ERROR:  input of anonymous composite types is not implemented
```

There is a test case like this in the `src/test/regress/sql/gporca.sql` test, but I'm not sure why our current pg_upgrade tests are not choking on that.

**Comment by hlinnaka** [comment]
> There is a test case like this in the src/test/regress/sql/gporca.sql test, but I'm not sure why our current pg_upgrade tests are not choking on that.

There's a `DROP SCHEMA orca CASCADE` at the end of the test file. That explains it; the object is not kept around, so it's not tested by pg_upgrade.

**Comment by kainwen** [comment]
After merging pgv12, the master branch now does not have the issue.

**Comment by dpandhi-git** [comment]
@jimmyyih can you check recommendation from Heiki for upgrade tests

**Comment by jimmyyih** [comment]
Just to make sure this Github issue doesn't get confused with https://github.com/greenplum-db/gpdb/issues/14941 or https://github.com/greenplum-db/gpdb/issues/14956 (both related but need their own separate fixes), I'll post my current findings here.

The reason why pg_restore fails even though the SQL looks correct is because custom types with non-catalog operators depend on search_path during the CREATE TABLE to find the correct operators for the custom type.  The binary upgrade SQL dumps clear the search path at the beginning so the CREATE TABLE will fail because it can't find the non-catalog operator.  A more simpler repro would be something like the following:
```
CREATE TYPE employee_type AS (empid int, empname text);

CREATE FUNCTION emp_equal(employee_type, employee_type) RETURNS boolean
  AS 'select $1.empid = $2.empid;'
  LANGUAGE SQL IMMUTABLE
  RETURNS null ON NULL INPUT;

CREATE SCHEMA fooschema;

CREATE OPERATOR fooschema.= (
        leftarg = employee_type,
        rightarg = employee_type,
        procedure = emp_equal
);

CREATE OPERATOR CLASS employee_incomplete_op_class default FOR TYPE employee_type
  USING btree AS
  operator 3 fooschema.=;

-- this CREATE TABLE will fail because fooschema is not in the search_path
CREATE TABLE employee_table (userid int, emp employee_type)
DISTRIBUTED BY (userid) PARTITION BY LIST(emp)
(
  PARTITION part1 VALUES('(1," ''foo''")')
);
```

Current attempt is to fix the issue at the source which seems to be in `make_partition_rules` where the operators are created with `makeSimpleA_Expr()` instead of `makeA_Expr()` which would allow us to fully qualify the operator and bypass the search_path dependency later on.

**Comment by jimmyyih** [comment]
Fixed in 6X_STABLE with commit d18208cf0bb82e2ba78121547f44f9615196744e.
